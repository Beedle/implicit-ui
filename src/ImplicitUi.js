import React from 'react';
import PropTypes from 'prop-types';

export default function ImplicitUi({components, ui}) {
  if (!components) return null;
  if (!ui) return null;

  return (
    ui.map((element, index) => {
      const UiElement = getElement(components, element);
      const props = getElementProps(element);

      if (!UiElement) return null;

      return <UiElement key={index} {...props} />;
    })
  );
}

ImplicitUi.propTypes = {ui: PropTypes.array.isRequired};

function getElement(components, element) {
  if (typeof element === 'string') return components[element];
  if (typeof element === 'object') return components[element.name];

  return undefined;
}

function getElementProps(element) {
  if (typeof element === 'string') return {};
  if (!element.props) return {};

  return element.props;
}

/*

const ONLY_ONE_ELEMENT = 1;
const FIRST_ELEMENT = 0;

function ImplicitElement({element: {name, childElements, props}, params}) {
  const ElementComp = getComponentByName(name);
  const parsedProps = parseProps(props);

  if (!childElements) return <ElementComp {...parsedProps} {...params} />;

  let childComponents = childElements.map(
    (childComp, index) =>
      <ImplicitElement key={index} element={childComp} params={params} />
  );

  if (childComponents.length === ONLY_ONE_ELEMENT) childComponents = childComponents[FIRST_ELEMENT];

  return <ElementComp {...parsedProps} {...params}>{childComponents}</ElementComp>;
}

ImplicitElement.propTypes = {
  element: PropTypes.object.isRequired,
  params: PropTypes.object
};

ImplicitElement.defaultProps = {params: {}};

function getComponentByName(name) {
  return componentList[name];
}

function parseProps(props) {
  if (!props) return props;

  const newProps = {...props};

  for (const key in newProps) {
    const obj = newProps[key];

    if (obj && Object.prototype.hasOwnProperty.call(obj, 'name')) newProps[key] = getComponentByName(obj.name);
  }

  return newProps;
} */
