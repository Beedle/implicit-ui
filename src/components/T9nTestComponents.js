import React from 'react';
import PropTypes from 'prop-types';

import {withT9n} from '../T9n';

export function Paragraph({text}) {
  return <p>{text}</p>;
}

Paragraph.propTypes = {text: PropTypes.string.isRequired};

export function ShowContext({children, t9nContext}) {
  if (!children) return <span>{t9nContext}</span>;

  return (
    <div>
      <span>{t9nContext}</span>
      <div>
        {children}
      </div>
    </div>
  );
}

ShowContext.propTypes = {
  t9nContext: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ])
};

ShowContext.defaultProps = {children: undefined};

export const SimpleTag = withT9n(({tagName}) => <span>{tagName}</span>, {tagName: '$simpleTag'});
