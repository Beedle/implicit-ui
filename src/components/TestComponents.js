import React from 'react';
import PropTypes from 'prop-types';

export function Paragraph({text}) {
  return <p>{text}</p>;
}

Paragraph.propTypes = {text: PropTypes.string.isRequired};

export function MyTitle() {
  return <h1>My Title</h1>;
}
