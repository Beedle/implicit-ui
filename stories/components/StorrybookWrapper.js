import React from 'react';

import './StorybookWrapper.scss'; //eslint-disable-line

export default function StorybookWrapper({children}) {
  return (
    <div className="storybook-wrapper">{children}</div>
  );
}
