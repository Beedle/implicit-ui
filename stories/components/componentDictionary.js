import React from 'react';

export function MyParagraph() {
  return <p>This is my Paragraph</p>;
}

export function MyTextParagraph({text}) {
  return <p>{text}</p>;
}

export function Container({children, text}) {
  return (
    <div>
      <h3>{text}</h3>
      <div>
        {children}
      </div>
    </div>
  );
}

export function PropsContainer({Comp1, Comp2}) {
  return (
    <div>
      <Comp1 />
      <Comp2 />
    </div>
  );
}
