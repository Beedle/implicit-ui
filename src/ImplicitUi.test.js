import React from 'react';
import renderer from 'react-test-renderer';

import ImplicitUi from './ImplicitUi';
import * as compCollection from './components/TestComponents';


describe('ImplicitUi', () => {
  it('should use strings as component name', () => {
    const uiConfig = [
      'MyTitle'
    ];

    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should use name prop as component name', () => {
    const uiConfig = [
      {name: 'MyTitle'}
    ];

    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render list of components', () => {
    const uiConfig = [
      'MyTitle',
      'MyTitle'
    ];

    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should pass prop property as props to component', () => {
    const uiConfig = [
      {
        name: 'Paragraph',
        props: {text: 'This is my text'}
      }
    ];

    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render children prop as children to component', () => {
    const uiConfig = [
      {
        name: 'Container',
        props: {text: 'MyHeader'},
        children: [
          'MyTitle',
          'MyTitle'
        ]
      }
    ];

    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
