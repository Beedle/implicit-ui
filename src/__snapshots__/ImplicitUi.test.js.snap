// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImplicitUi should pass prop property as props to component 1`] = `
<p>
  This is my text
</p>
`;

exports[`ImplicitUi should render list of components 1`] = `
Array [
  <h1>
    My Title
  </h1>,
  <h1>
    My Title
  </h1>,
]
`;

exports[`ImplicitUi should use name prop as component name 1`] = `
<h1>
  My Title
</h1>
`;

exports[`ImplicitUi should use strings as component name 1`] = `
<h1>
  My Title
</h1>
`;
